tests/keypair_test.go

```diff
@@ -0,0 +1,250 @@
+	var n int64
+	keystore.SerialGenerator = func() *big.Int {
+		n++
+		return big.NewInt(n)
+	}
+
+	// We define a function so we can rebuild the tasks, because we modify in-place when running
+	buildTasks := func() map[string]fi.Task {
+		ca := &fitasks.Keypair{
+			Name:      fi.String(fi.CertificateId_CA),
+			Lifecycle: &lifecycle,
+			Subject:   "cn=kubernetes",
+			Type:      "ca",
+		}
+
+		kubelet := &fitasks.Keypair{
+			Name:      fi.String("kubelet"),
+			Lifecycle: &lifecycle,
+
+			Subject: "o=nodes,cn=kubelet",
+			Type:    "client",
+			Signer:  ca,
+		}
+
+		tasks := make(map[string]fi.Task)
+		tasks["ca"] = ca
+		tasks["kubelet"] = kubelet
+		return tasks
+	}
+
+	t.Logf("Building some keypairs")
+	{
+		allTasks := buildTasks()
+
+		context, err := fi.NewContext(target, nil, nil, keystore, nil, nil, true, allTasks)
+		if err != nil {
+			t.Fatalf("error building context: %v", err)
+		}
+
+		if err := context.RunTasks(defaultDeadline); err != nil {
+			t.Fatalf("unexpected error during Run: %v", err)
+		}
+	}
+
+	// Check that the expected files were generated
+	expected := []string{
+		"memfs://keystore/issued/ca/1.crt",
+		"memfs://keystore/issued/ca/keyset.yaml",
+		"memfs://keystore/issued/kubelet/2.crt",
+		"memfs://keystore/issued/kubelet/keyset.yaml",
+		"memfs://keystore/private/ca/1.key",
+		"memfs://keystore/private/ca/keyset.yaml",
+		"memfs://keystore/private/kubelet/2.key",
+		"memfs://keystore/private/kubelet/keyset.yaml",
+	}
+	checkPaths(t, basedir, expected)
+
+	// Save the contents of those files
+	contents := make(map[string][]byte)
+	for _, k := range expected {
+		p, err := vfs.Context.BuildVfsPath(k)
+		if err != nil {
+			t.Fatalf("error building vfs path: %v", err)
+		}
+		b, err := p.ReadFile()
+		if err != nil {
+			t.Fatalf("error reading vfs path: %v", err)
+		}
+		contents[k] = b
+	}
+
+	t.Logf("verifying that rerunning tasks does not change keys")
+	{
+		allTasks := buildTasks()
+
+		context, err := fi.NewContext(target, nil, nil, keystore, nil, nil, true, allTasks)
+		if err != nil {
+			t.Fatalf("error building context: %v", err)
+		}
+
+		if err := context.RunTasks(defaultDeadline); err != nil {
+			t.Fatalf("unexpected error during Run: %v", err)
+		}
+	}
+	checkContents(t, basedir, contents)
+
+	t.Logf("deleting keyset.yaml files and verifying they are recreated")
+	FailOnError(t, basedir.Join("issued/ca/keyset.yaml").Remove())
+	FailOnError(t, basedir.Join("issued/kubelet/keyset.yaml").Remove())
+	FailOnError(t, basedir.Join("private/ca/keyset.yaml").Remove())
+	FailOnError(t, basedir.Join("private/kubelet/keyset.yaml").Remove())
+
+	{
+		allTasks := buildTasks()
+
+		context, err := fi.NewContext(target, nil, nil, keystore, nil, nil, true, allTasks)
+		if err != nil {
+			t.Fatalf("error building context: %v", err)
+		}
+
+		if err := context.RunTasks(defaultDeadline); err != nil {
+			t.Fatalf("unexpected error during Run: %v", err)
+		}
+	}
+	checkContents(t, basedir, contents)
```

Comment:
With my changes this seems to be always failing.  Not certain if I do not understand the test or what.  

https://github.com/chrislovecnm/kops/blob/903b153cdcb57a3495001996c832a7c5b92da381/upup/pkg/fi/fitasks/keypair.go#L179

Is always passing the if statement.

Reply:
Here are the logs

```
I0308 21:54:25.601958   58069 topological_sort.go:63] Dependencies:
I0308 21:54:25.601982   58069 topological_sort.go:65] 	ca:	[]
I0308 21:54:25.602010   58069 topological_sort.go:65] 	kubelet:	[ca]
I0308 21:54:25.602022   58069 executor.go:91] Tasks: 0 done / 2 total; 1 can run
I0308 21:54:25.602043   58069 executor.go:166] Executing task "ca": *fitasks.Keypair {"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"}
I0308 21:54:25.602633   58069 keypair.go:181] creating brand new certificate
I0308 21:54:25.602648   58069 keypair.go:199] Creating PKI keypair "ca"
I0308 21:54:25.602664   58069 keypair.go:210] Creating privateKey "ca"
I0308 21:54:25.862300   58069 vfs_castore.go:725] Issuing new certificate: "ca"
I0308 21:54:25.866478   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:25.866543   58069 keypair.go:228] created certificate &{{[] [] [] [] [] [] []  kubernetes [{2.5.4.3 kubernetes}] []} true 0xc420414000 0xc420332400}
I0308 21:54:25.866580   58069 executor.go:91] Tasks: 1 done / 2 total; 1 can run
I0308 21:54:25.866601   58069 executor.go:166] Executing task "kubelet": *fitasks.Keypair {"Name":"kubelet","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":{"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"},"subject":"o=nodes,cn=kubelet","type":"client","keypairType":"Keypair"}
I0308 21:54:25.866697   58069 keypair.go:181] creating brand new certificate
I0308 21:54:25.866704   58069 keypair.go:199] Creating PKI keypair "kubelet"
I0308 21:54:25.866711   58069 keypair.go:210] Creating privateKey "kubelet"
I0308 21:54:26.128270   58069 vfs_castore.go:725] Issuing new certificate: "kubelet"
I0308 21:54:26.128802   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.129454   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.129567   58069 privatekey.go:169] Parsing pem block: "RSA PRIVATE KEY"
I0308 21:54:26.132818   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.132889   58069 keypair.go:228] created certificate &{{[] [nodes] [] [] [] [] []  kubelet [{2.5.4.10 nodes} {2.5.4.3 kubelet}] []} false 0xc4201d4000 0xc420491c90}
I0308 21:54:26.132947   58069 executor.go:91] Tasks: 2 done / 2 total; 0 can run
I0308 21:54:26.133090   58069 topological_sort.go:63] Dependencies:
I0308 21:54:26.133096   58069 topological_sort.go:65] 	ca:	[]
I0308 21:54:26.133101   58069 topological_sort.go:65] 	kubelet:	[ca]
I0308 21:54:26.133126   58069 executor.go:91] Tasks: 0 done / 2 total; 1 can run
I0308 21:54:26.133143   58069 executor.go:166] Executing task "ca": *fitasks.Keypair {"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"}
I0308 21:54:26.133324   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.133403   58069 executor.go:91] Tasks: 1 done / 2 total; 1 can run
I0308 21:54:26.133415   58069 executor.go:166] Executing task "kubelet": *fitasks.Keypair {"Name":"kubelet","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":{"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"},"subject":"o=nodes,cn=kubelet","type":"client","keypairType":"Keypair"}
I0308 21:54:26.133568   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.133812   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.133878   58069 privatekey.go:169] Parsing pem block: "RSA PRIVATE KEY"
I0308 21:54:26.134095   58069 executor.go:91] Tasks: 2 done / 2 total; 0 can run
I0308 21:54:26.134236   58069 topological_sort.go:63] Dependencies:
I0308 21:54:26.134242   58069 topological_sort.go:65] 	kubelet:	[ca]
I0308 21:54:26.134246   58069 topological_sort.go:65] 	ca:	[]
I0308 21:54:26.134271   58069 executor.go:91] Tasks: 0 done / 2 total; 1 can run
I0308 21:54:26.134283   58069 executor.go:166] Executing task "ca": *fitasks.Keypair {"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"}
I0308 21:54:26.134329   58069 keypair.go:181] creating brand new certificate
I0308 21:54:26.134335   58069 keypair.go:199] Creating PKI keypair "ca"
I0308 21:54:26.134341   58069 vfs_castore.go:725] Issuing new certificate: "ca"
I0308 21:54:26.136580   58069 privatekey.go:169] Parsing pem block: "RSA PRIVATE KEY"
I0308 21:54:26.137295   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.137601   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.137656   58069 keypair.go:228] created certificate &{{[] [] [] [] [] [] []  kubernetes [{2.5.4.3 kubernetes}] []} true 0xc420200500 0xc4202ff9d0}
I0308 21:54:26.137676   58069 executor.go:91] Tasks: 1 done / 2 total; 1 can run
I0308 21:54:26.137691   58069 executor.go:166] Executing task "kubelet": *fitasks.Keypair {"Name":"kubelet","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":{"Name":"ca","alternateNames":null,"alternateNameTasks":null,"Lifecycle":"Sync","Signer":null,"subject":"cn=kubernetes","type":"ca","keypairType":"Keypair"},"subject":"o=nodes,cn=kubelet","type":"client","keypairType":"Keypair"}
I0308 21:54:26.137758   58069 keypair.go:181] creating brand new certificate
I0308 21:54:26.137780   58069 keypair.go:199] Creating PKI keypair "kubelet"
I0308 21:54:26.137787   58069 keypair.go:210] Creating privateKey "kubelet"
I0308 21:54:26.245911   58069 vfs_castore.go:725] Issuing new certificate: "kubelet"
I0308 21:54:26.248417   58069 privatekey.go:169] Parsing pem block: "RSA PRIVATE KEY"
I0308 21:54:26.249990   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.250408   58069 certificate.go:103] Parsing pem block: "CERTIFICATE"
I0308 21:54:26.250514   58069 keypair.go:228] created certificate &{{[] [nodes] [] [] [] [] []  kubelet [{2.5.4.10 nodes} {2.5.4.3 kubelet}] []} false 0xc420415900 0xc42028b6a0}
I0308 21:54:26.250572   58069 executor.go:91] Tasks: 2 done / 2 total; 0 can run
--- FAIL: TestKeypairUpgrade (0.65s)
	keypair_test.go:114: Building some keypairs
	keypair_test.go:168: verifying that rerunning tasks does not change keys
	keypair_test.go:183: deleting keyset.yaml files and verifying they are recreated
	keypair_test.go:257: unexpected paths: [memfs://keystore/issued/ca/1.crt memfs://keystore/issued/ca/3.crt memfs://keystore/issued/ca/keyset.yaml memfs://keystore/issued/kubelet/2.crt memfs://keystore/issued/kubelet/4.crt memfs://keystore/issued/kubelet/keyset.yaml memfs://keystore/private/ca/1.key memfs://keystore/private/ca/3.key memfs://keystore/private/ca/keyset.yaml memfs://keystore/private/kubelet/2.key memfs://keystore/private/kubelet/4.key memfs://keystore/private/kubelet/keyset.yaml]
```

We are recreating the certs all the time.